cloudbuild_public.yaml

```diff
@@ -0,0 +1,70 @@
+steps:
+- name: 'gcr.io/cloud-builders/docker'
+  entrypoint: 'bash'
+  args: [
+    '-c',
+    'docker pull us.gcr.io/$PROJECT_ID/$REPO_NAME:nightly || true',
+  ]
+- name: 'gcr.io/cloud-builders/docker'
+  args: 
+    [
+      'build',
+      '-t', 'us.gcr.io/$PROJECT_ID/$REPO_NAME:$COMMIT_SHA',
+      '-t', 'us.gcr.io/$PROJECT_ID/$REPO_NAME:nightly',
+      '--cache-from', 'us.gcr.io/$PROJECT_ID/$REPO_NAME:nightly',
+      '--cache-from', 'us.gcr.io/$PROJECT_ID/$REPO_NAME:$COMMIT_SHA',
+      '.'
+    ]
+# CDC tests
+- name: 'gcr.io/$PROJECT_ID/docker-compose'
```

Comment:
Is it ok run the build step of the cdc image when running docker compose or should I build a development image and use that one? The standard image built above cannot run the unit test because the target is different.

Reply:
I don't think we should run unit tests inside the docker image we build. The tests we perform on Docker images are e2e tests to ensure the image actually works. Anything else should be covered in more specific test workflows in GitHub actions.

Unless you pass the `--target` parameter to Docker, it will end up building the dev/test image all the time with the additional dependencies. An alternative to this is using a separate Dockerfile for tests, such as `dockerfile.test`. This is also what we have been discussing with @evanh for Snuba repo as right now we are shipping dev dependencies on prod/self-hosted images.

Reply:
Right, I should change the way the docker image is built.

About the tests. If I add an integration test, should I then run it with github actions as well or just in cloudbuild ?

Reply:
I added a single end to end test that sets everything up with docker compose. That is the only test that runs on cloud build now.

Reply:
>I added a single end to end test that sets everything up with docker compose. That is the only test that runs on cloud build now.

Exactly what I was thinking.

>About the tests. If I add an integration test, should I then run it with github actions as well or just in cloudbuild ?

Depends on how you define an integration test. The way we did in the past is:

- Any "programmatic test" that requires access to source code should be done through GitHub actions, without the image.
- The built image then should be tested as a black box using e2e tests to ensure the packaging and image build is sane (but not all the functionality as it should already be tested with the aforementioned unit and integration tests).

This isolates the image building step, makes it simpler (you don't need to think about a test image), and avoids duplicated test runs. That said it introduces a risk vector where we don't actually fully test the end production image. If this is a concern, we may think of alternatives, including a test image build.

Reply:
Ok, then I think this way should be enough.
- GHA runs the unit tests.
- this one uses the production image it builds to run it and does not build a dedicated one.

I am ok in not running this test on GHA as anyway cloudbuild would be running it anyway before generating the image.  There is no scenario we can have a new image that was created without running all tests.